website/src/documentation/patterns/custom-io-patterns.md

```diff
@@ -0,0 +1,41 @@
+---
+layout: section
+title: "Custom I/O patterns"
+section_menu: section-menu/documentation.html
+permalink: /documentation/patterns/custom-io-patterns/
+---
+<!--
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Custom I/O patterns
+
+This page describes common patterns in pipelines with [custom I/O connectors](https://beam.apache.org/documentation/io/developing-io-overview/). Custom I/O connectors connect pipelines to databases that aren't supported by Beam's [built-in I/O transforms](https://beam.apache.org/documentation/io/built-in/).
+
+<nav class="language-switcher">
+  <strong>Adapt for:</strong>
+  <ul>
+    <li data-type="language-java" class="active">Java SDK</li>
+    <li data-type="language-py">Python SDK</li>
+  </ul>
+</nav>
+
+## Choosing between built-in and custom connectors
+
+[Built-in I/O connectors]({{ site.baseurl }}//documentation/io/built-in/) are tested and hardened, so use them whenever possible. Only use custom I/O connectors when:
+* No built-in options exist
+* Your pipeline pulls in a small subset of source data
+
+For instance, use a custom I/O connector to enrich pipeline elements with a small subset of source data. If you’re processing a sales order and adding information to each purchase, you can use a custom I/O connector to pull the small subset of data into your pipeline (instead of processing the entire source).
+
+Beam distributes work across many threads, so custom I/O connectors can increase your data source’s load average. You can reduce the load with the <span class="language-java">[start]({{ site.baseurl }}/releases/javadoc/2.9.0/org/apache/beam/sdk/transforms/DoFn.StartBundle.html)</span><span class="language-py">[start]({{ site.baseurl }}/releases/pydoc/2.9.0/apache_beam.transforms.core.html?highlight=bundle#apache_beam.transforms.core.DoFn.start_bundle)</span> and <span class="language-java">[finish]({{ site.baseurl }}/releases/javadoc/2.9.0/org/apache/beam/sdk/transforms/DoFn.FinishBundle.html)</span><span class="language-py">[finish]({{ site.baseurl }}/releases/pydoc/2.9.0/apache_beam.transforms.core.html?highlight=bundle#apache_beam.transforms.core.DoFn.finish_bundle)</span> bundle annotations.
```

Comment:
for javadoc/pydoc links, you can use "current" instead of the version number (such as "2.9.0") in the URL. that way it will automatically redirect to the most recent version.

the other gotcha about javadoc/pydoc links are that you can't use site.baseurl due to how the website is set up (they aren't part of the jekyll generated pages). so you'll have to actually use https://beam.apache.org in any links to javadoc/pydoc pages

you can view link/html issues such as these in the test results by going to the test results section at the bottom of the PR, clicking the Details link to the right of "Website ("Run Website PreCommit") — FAILURE", and then looking at the Console Output link on the left side of the Jenkins results page.  (it looks like there might be a link issue in the overview page also)